### DISCUSSION TITLE
Add `--replace` flag to `move`

### DISCUSSION THREAD

**Discussion by pokey** [arxanas/git-branchless#553] [Ideas]
Would be a substitute for `--dest`, and would accept a revset.  Would be nice to support on https://github.com/arxanas/git-branchless/discussions/551 as well

With this discussion and https://github.com/arxanas/git-branchless/discussions/551, I'm thinking of a workflow where you can keep one branch which has your working copy with all the changes, and then duplicate the changes in one or more branches for review.  These commands could enable a workflow for propagating changes back and forth between the branches by rewriting the sections as they change

**Comment by arxanas** [comment]
You might also want to check out [git-branchstack](https://git.sr.ht/~krobelus/git-branchstack) by @krobelus.

Generally, in git-branchless, I don't separate the stacks out; instead, if I want to make changes to a previous commit, I just check it out and start editing it (and either run `git amend` to rebase the descendant commits, or `git commit` to defer rebasing until later).

It might also be desirable to have only one view locally and generate the individual branches for review synthetically as part of `git submit`.

You could probably also approximate this with something like `git move -x def456 -d abc123^ && git move -s 'children(abc123)' && git hide abc123`.

**Reply by arxanas** [reply]
Oh, it's also worth checking out [git-assembler](https://www.thregr.org/~wavexx/software/git-assembler/#testing-two-branches-together). The workflows are kind of the opposite. In git-branchstack, you build one big branch and split it up afterwards; in git-assembler, you write small branches and stack them together after the fact.

**Reply by pokey** [reply]
Ah cool; git-assembler looks nice.  

And yeah had the thought that this could be done in two steps with a drop followed by an insertion.  Although I'm not sure your proposal is quite right.  Presumably `def456` represents the source revset, and `abc123` represents the revset to be replaced. In this case, the first step will attach `def456` to the parent of `abc123`, but then the next command is assuming that we have checked out the tip of the newly moved commits.  Unfortunately, I'm not sure I see a reliable way to automatically target the tip of the newly moved commits as the destination of that second move 🤔

Maybe `move` could have some kind of `--raw` mode where it outputs topologically sorted list of moved commits

I think a `--replace` flag could handle this precisely, and could do things like ensuring that

- the destination revset has only one root,
- if the destination revset has children, then the source revset has only one head, as otherwise it's unclear what to do with the children of the destination revset

**Reply by pokey** [reply]
I guess another disadvantage of a two-step process is that I'd like for the whole thing to be done speculatively in memory, and to bail out if either the drop or the insertion will result in a conflict

I wonder if it would make sense to create some kind of simple language for a sequence of actions that should all be done speculatively in memory and fail if any of them results in a conflict.  Like eg

```
git execute < steps.txt
```

Where `steps.txt` is

```
move -x def456 -d abc123^
move -s 'children(abc123)'
hide abc123
```

It would be cool if there could be assignment there too, so eg 

```
new_commits = move -x def456 -d abc123^
move -s 'children(abc123)' -d '$new_commits'
hide abc123
```

**Reply by pokey** [reply]
It would also be neat to have a configuration language like git-assembler, and even to make it two-way, and could auto update on merge.  A lot of that could just be built on top of git-branchless tbh

**Reply by arxanas** [reply]
> Although I'm not sure your proposal is quite right

Oops, I think you're right.

> It would also be neat to have a configuration language like git-assembler, and even to make it two-way, and could auto update on merge. A lot of that could just be built on top of git-branchless tbh

It might be more straightforward to extend the existing revset language to be able to do side-effectful computations, i.e. just write something like

```
new_commits = move!(def456, abc123);
move!(children(abc123):, new_commits);
hide!(abc123)
```

It's perhaps worth noting that several git-branchless operations could be expressed in this way already. `git sync` is more or less equivalent to something like `fetch!(main); move!(draft(), main)` (except that it tries to move stacks individually, so that the failure of one doesn't cause the failure of others) and `git restack` would be something like (with made-up syntax) `foreach(hidden():, x => move!((children(x) - hidden()):, successor(x))`.

cc @martinvonz: for jj, we once talked about "what if instead of a main branch, we had a main revset?", for which we decided that wouldn't work for some cases, but what if we generalized it *even further* and had side-effectful revset scripts?

**Reply by pokey** [reply]
Yes I think a language like that would be awesome

I still think `--replace` is a useful flag as well, though

**Reply by pokey** [reply]
Re the revset language extension: would it be done speculatively and bail out entirely if there would be a conflict?

**Reply by arxanas** [reply]
I would envision it being done speculatively. There's probably more than one value on the spectrum of speculation. For example, for `git sync`, we essentially want to do "as much as possible and leave the rest", which is somewhere between "accept no conflicts" and "resolve all conflicts now". So ideally we could choose from any one of those three behaviors.
